src/common/src/main/scala/osmesa/common/model/Change.scala

```diff
@@ -25,55 +28,64 @@ case class Change(id: Long,
                   sequence: Int)
 
 object Change {
+  def empty: Change = Change(-1, "", Map.empty, None, None, None, None, -1, "1970-01-01 00:00:00".asInstanceOf[Timestamp], -1, "", -1, false, -1)
+
   implicit def stringToTimestamp(s: String): Timestamp =
     Timestamp.from(DateTime.parse(s).toDate.toInstant)
 
-  def fromXML(node: Node, action: Action, sequence: Int): Change = {
-    val `type` = node.label
-    val id = (node \@ "id").toLong
-    val tags =
-      (node \ "tag").map(tag => (tag \@ "k", tag \@ "v")).toMap
-    val lat = node \@ "lat" match {
-      case "" => None
-      case v  => Some(v.toDouble)
-    }
-    val lon = node \@ "lon" match {
-      case "" => None
-      case v  => Some(v.toDouble)
-    }
-    val nds = `type` match {
-      case "way" =>
-        Some((node \ "nd").map(Nd.fromXML))
-      case _ => None
+  class ChangeHandler(sequence: Int) extends DefaultHandler {
+    val changeSeq = Queue.empty[Change]
+    var action: Actions.Action = Actions.Delete
+    var working: Change = null
+    override def startElement(uri: String, localName: String, qName: String, attributes: sax.Attributes) = {
+      val attrs =
+        (for {
+           i <- Range(0, attributes.getLength).toSeq
+         } yield (attributes.getQName(i) -> attributes.getValue(i))).toMap
+      qName.toLowerCase match {
+        case label if Set("create", "delete", "modify").contains(label) =>
+          action = Actions.fromString(qName)
+        case label if Set("node", "way", "relation").contains(label) =>
```

Comment:
Extract the `Set` to a constant to avoid recreating in each `startElement` call?